pkg/reconciler/pipeline/controller.go

```diff
@@ -435,6 +435,12 @@ func buildVertices(pl *dfv1.Pipeline) map[string]dfv1.Vertex {
 		replicas := int32(1)
 		if pl.Status.Phase == dfv1.PipelinePhasePaused {
 			replicas = int32(0)
+		} else if v.UDF != nil && v.UDF.GroupBy != nil {
```

Comment:
@dpadhiar Hey Dillen, I see we are looping through the fromEdges, does it mean if all fromEdges have parallelism, we will always take the last edge's number? I am not quite familiar with reduce replica but that doesn't sound right to me intuitively.

Also two nitpickings: 1. I think you can use `v.IsReduceUDF() instead of `v.UDF != nil && v.UDF.GroupBy != nil` and 2. could you add a unit test case to cover this change?

Reply:
@KeranYang Hi Keran, thanks for taking a look. For `v.IsReduceUDF()`, I also tried to use this but since `v` is a `dfv1.AbstractVertex`, I couldn't use this function but I did use the logic outright to check for a reduce vertex.

You are correct that looping through all the from edges and taking the last `Parallelism` value is incorrect.  I am still getting familiar with Numa so I have not seen a UDF with multiple from edges that have different `Parallelism` values. If this was the case, @whynowy is the largest `Parallelism` value used?

As for unit test, this PR is closed but I will open another to include that after we confirm the above.

Reply:
The reason I dismissed the for loop logic is, there's only 1 from edge for a vertex, technically we should only take the 1st item of the array to make logic look right. @dpadhiar - please address this together with the unit test.

Reply:
Will take care of this now, should be a quick turnaround.